### PR TITLE
feat: limit date picker range

### PR DESCRIPTION
**Pull request by YiniXu9506**
**Only take effect on cloud side.**

- support quick time range: 2 days, 7 days, 15 days.
- disable dates which before 15 days ago and after today.
- disable time which before current hour on 15 days ago, and after current hour today.
- hide zoomOut feature for timeSelector on cloud side
image
image
image

### CODE REVIEW

**Review by shhdgit — APPROVED** [review]

**Comment by baurine** [comment]
Let's remove the JIRA issue link because some people can't access it, we can attach the JIRA issue link in the clinic repo side PR and link to this PR.

File: ui/packages/tidb-dashboard-lib/src/components/TimeRangeSelector/index.tsx [hunk 1]
```diff
@@ -142,6 +147,105 @@ function TimeRangeSelector({
...
+    } else if (Math.abs(size) < 86400) {
+      // Less than one day, divide in hours
+      return toFixedScaled(size / 3600, decimals, scaledDecimals, 4, ' hour')
+    } else {
+      // Less than one week, divide in days
+      return toFixedScaled(size / 86400, decimals, scaledDecimals, 5, ' day')
+    }
+  }
+
+  // eslint-disable-next-line arrow-body-style
```
Comment (baurine) [inline ui/packages/tidb-dashboard-lib/src/components/TimeRangeSelector/index.tsx:215]: Why this function needs this comment while others don't need?

File: ui/packages/tidb-dashboard-lib/src/components/TimeRangeSelector/index.tsx [hunk 2]
```diff
@@ -142,6 +147,105 @@ function TimeRangeSelector({
...
+  const disabledDate = (current) => {
+    const today = dayjs().startOf('hour')
+    // Can not select days before 15 days ago
+    const tooEarly =
+      today.subtract(15 * 24, 'hour') > dayjs(current).startOf('hour')
```
Comment (baurine) [inline ui/packages/tidb-dashboard-lib/src/components/TimeRangeSelector/index.tsx:220]: I think the `15` should not be hard-code, it should be calculated from `recent_seconds` option.

File: ui/packages/tidb-dashboard-lib/src/components/TimeRangeSelector/index.tsx [hunk 3]
```diff
@@ -142,6 +147,105 @@ function TimeRangeSelector({
...
+    const hour = dayjs().hour()
+    // is current day
+    if (current && current.isSame(dayjs(), 'day')) {
+      return {
+        disabledHours: () => hours.slice(hour)
+      }
+    }
+
+    // is 15 day ago
+    if (current && current.isSame(dayjs().subtract(15, 'day'), 'day')) {
```
Comment (baurine) [inline ui/packages/tidb-dashboard-lib/src/components/TimeRangeSelector/index.tsx:240]: ditto

**Comment by YiniXu9506** [comment]
> Let's remove the JIRA issue link because external people can't access it, we can attach the JIRA issue link in the clinic repo side PR and link to this PR.

Got it.

**Comment by baurine** [comment]
Did we miss the package.json file change?

File: ui/packages/tidb-dashboard-lib/src/components/TimeRangeSelector/index.tsx (hunk 1 above)
Comment (YiniXu9506) [reply ui/packages/tidb-dashboard-lib/src/components/TimeRangeSelector/index.tsx:215]: Actually we don't need this. I got this comment due to some plugin of vscode. I will remove it.

File: ui/packages/tidb-dashboard-lib/src/components/TimeRangeSelector/index.tsx [hunk 4]
```diff
@@ -142,6 +147,112 @@ function TimeRangeSelector({
...
+    if (
+      value1 !== null &&
+      value1 !== undefined &&
+      value2 !== null &&
+      value2 !== undefined
+    ) {
+      return value1 - value2
+    }
+    return undefined
+  }
```
Comment (baurine) [inline ui/packages/tidb-dashboard-lib/src/components/TimeRangeSelector/index.tsx:160]: This method can be removed out of the component

File: ui/packages/tidb-dashboard-lib/src/components/TimeRangeSelector/index.tsx [hunk 5]
```diff
@@ -142,6 +147,112 @@ function TimeRangeSelector({
...
+
+  const customValueFormat = (
+    size: number,
+    decimals?: DecimalCount,
+    scaledDecimals?: DecimalCount
+  ) => {
```
Comment (baurine) [inline ui/packages/tidb-dashboard-lib/src/components/TimeRangeSelector/index.tsx:166]: This method can be removed out of the component

File: ui/packages/tidb-dashboard-lib/src/components/TimeRangeSelector/index.tsx [hunk 6]
```diff
@@ -142,6 +147,112 @@ function TimeRangeSelector({
...
+    // Can not select days before 15 days ago
+    const tooEarly =
+      today.subtract(selectableHours, 'hour') > dayjs(current).startOf('hour')
+    // Can not select days after today
+    const tooLate = today < dayjs(current).startOf('hour')
+    return current && (tooEarly || tooLate)
+  }
+
+  // array of 24 numbers, start from 0
+  const hours = [...Array(24).keys()]
```
Comment (baurine) [inline ui/packages/tidb-dashboard-lib/src/components/TimeRangeSelector/index.tsx:231]: This var can be removed out of the component

**Comment by YiniXu9506** [comment]
Holding this PR and waiting for notification.

**Review by shhdgit — APPROVED** [review]

**Pull request merged**